### DISCUSSION TITLE
How to return HTML to support HTMX

### DISCUSSION THREAD

**Discussion by bartonhammond** [benallfree/pocketpages#17] [Q&A]
On my form to create a todo, I have:
```
<form hx-post="/newtodo" hx-trigger="submit" hx-target=".todo-list" hx-swap="afterbegin" hx-on::after-request="this.reset()" hx-swap="none">
      <input name="todo" class="new-todo" placeholder="What needs to be done?" autofocus>
    </form>
```
This calls into route properly:
```
routerAdd("POST", "/newtodo", (c) => {
	const data = $apis.requestInfo(c).data;
	const collection = $app.dao().findCollectionByNameOrId("todos");
	const record = new Record(collection);
	const form = new RecordUpsertForm($app, record);
	form.loadData({
		field: data.todo,
		done: false,
	});
	form.submit();
});
```
But I think I want to return the HTML that contains the todo list which is this:

```
<ul class="todo-list">
  <% data.todos.forEach(function(todo){ %>
  <li class="<%= todo.done  ? 'completed' : '' %>">
    <div class="view">
      <input class="toggle" type="checkbox" <%= todo.done  ? 'checked' : '' %>>
      <label>
        <%= todo.field %>
      </label>
      <button class="destroy"></button>
    </div>
    <input class="edit" value="Create a TodoMVC template">
  </li>
  <% }); %>
</ul>
```

How can this be done?  I don't have access to EJS w/in the server side code.

**Comment by benallfree** [comment]
Instead of `routerAdd`, create a pocketpage like `api/newtodo.ejs` and handle it in there.

**Comment by bartonhammond** [comment]
Is there any documentation or example?  I don't understand what that ejs file would look like.

**Comment by benallfree** [comment]
https://pocketpages.dev/docs/routing#basic-example

**Comment by bartonhammond** [comment]
The file `newtodo.ejs` is located at `app/pages/api/newtodo.ejs`
```
console.log('newtodo');
<h1>Request Method: <%= ctx.method() %></h1>
<p>Request Path: <%= ctx.path() %></p>
```

The `index.ejs` has the form tag:
```
 <form hx-post="/api/newtodo" hx-trigger="submit" hx-target=".todo-list" hx-swap="afterbegin" hx-on::after-request="this.reset()" hx-swap="none">
      <input name="todo" class="new-todo" placeholder="What needs to be done?" autofocus>
    </form>
```

I get this error when submitting the form
```
htmx.org@2.0.2:1 Response Status Error Code 404 from /api/newtodo/
```

**Comment by bartonhammond** [comment]
I can not get the `api/newtodo` to work - it gets a 404 when submitting.

 If I put the file in `pages` directly it doesn't get a 404.  It gets a 405 - Method not allowed.
<img width="790" alt="Screenshot 2024-11-22 at 12 29 15 PM" src="https://github.com/user-attachments/assets/ab950763-e1c2-4cae-9534-8cdd5a36e249">

**Comment by benallfree** [comment]
I see there is a `/` at the end, I wonder if that has something to do with it. Can you browse directly to it and hit it?

**Comment by bartonhammond** [comment]
I've tried changing the various `/` - in the front, in the back, etc.

I can not access the page when it is w/in the `api`.  I can only access the page when it is in the root of `pages`.

When the `newtodo.ejs` is accessed as `http://127.0.0.1:8090/newtodo` and the page is this, it works
```
<h1>Hello</h1>
```

But if I update to this:
```
<h1>Hello</h1>
<h1>Request Method: <%= ctx.method() %></h1>
```
I get this:
```
PocketPages Error
TypeError: /newtodo.ejs:2
    1| <h1>Hello</h1>
 >> 2| <h1>Request Method: <%= ctx.method() %></h1>

Object has no member 'method'
	at anonymous (<eval>:12:36(32))
	at apply (native)
	at anonymous (node_modules/pocketpages/node_modules/pocketbase-ejs/lib/ejs.js:702:16(24))
	at node_modules/pocketpages/node_modules/pocketbase-ejs/lib/ejs.js:490:27(103)
	at renderFile (node_modules/pocketpages/src/main.ts:387:28(24))
	at node_modules/pocketpages/src/main.ts:419:31(387)
	at apply (native)
	at <eval>:3:35(8)
```

**Comment by benallfree** [comment]
I'll work on a fix.

**Comment by bartonhammond** [comment]
@benallfree Thanks!  Looking forward to it...

**Comment by bartonhammond** [comment]
If I change `api` to `foo` I can invoke but then still get the issue w/ ctx 
<img width="879" alt="Screenshot 2024-11-22 at 5 14 48 PM" src="https://github.com/user-attachments/assets/05532dfb-fa33-4039-9f21-b03c9419fd33">

**Comment by benallfree** [comment]
My mistake. `/api` is a reserved path and is forwarded straight to PocketBase. Use `xapi` or something like thiat.

v0.9.4 contains the updates you need, and I wrote an htmx tutorial and starter kit too.

https://pocketpages.dev/docs/htmx
https://github.com/benallfree/pocketpages/blob/main/starters/htmx/README.md

**Comment by benallfree** [comment]
`ctx.request().method`, right? 

https://pocketbase.io/jsvm/interfaces/http.Request.html#method

**Comment by bartonhammond** [comment]
Will pick up tomorrow - thanks for all.

**Comment by bartonhammond** [comment]
On the htmx docs, is that example app available?  It looks more complete then the starter template.

As to the `ctx.method()`, I saw that here: [https://pocketpages.dev/docs/request-context](https://pocketpages.dev/docs/request-context) so maybe those docs should be updated?

I've upgraded to v0.9.4 and restructured the directories and now the ejs file in `xapi` is getting called properly.

How do I get the `data` set from with in the xapi call?

This is the _partial

```
<% data.todos.forEach(function(todo){ %>
....
<% }); %>
```

The data is set from the `+load.js`.  If I want to reuse the partial, I need to set `data`.  Is there a way to invoke the `+load.js` file?  The logic in there is exact as the xapi. 
Here's the `+load.js`
```
module.exports = () => {
   const result = arrayOf(new DynamicModel({
        "id":     "",
        "field": "",
        "done":    false,
    }));

    $app.dao().db()
        .newQuery("SELECT id, field, done FROM todos LIMIT 100")
        .all(result) 
    
	return {
		todos: result
	};
};
```
Here's the part of the xapi post call that I'm stuck at: (notice it's the same as the `+load.js`

```
<%
......
	$app
		.dao()
		.db()
		.newQuery("SELECT id, field, done FROM todos LIMIT 100")
		.all(result); // throw an error on db failure

///////.  how get the data object set for the partial?
	ctx.data.todos = result;
//////
 
%>
<%- include('/_private/todolist.ejs') %>
```

**Comment by bartonhammond** [comment]
This works in the xapi code
```
     ....
    let data = {
        todos: result
    }

%>
<%- include('/_private/todolist.ejs', {data}) %>
```

So the last thing for this issue, how can I keep the `+load.js` DRY?

**Comment by benallfree** [comment]
Do you have a repo I can look at? I’m kind of confused by the code samples.

**Reply by bartonhammond** [reply]
https://github.com/bartonhammond/todomvc

Making progress.  Right now I'm working on clicking the "done" button.  The "xapi/toggle" gets called fine and the record is updated in the database.  But the output from the template is messed up - parts of it are encoded (I think?) and the "field" value is missing.  Not sure why.

**Comment by benallfree** [comment]
Ok so those db calls return a [model.Record](https://pocketbase.io/jsvm/interfaces/models.Record.html) object which is something like `rec.get('field')` not `rec.field`.

Or you can do `const plainObject = JSON.parse(JSON.stringify(rec))` and then do `plainObject.field` should work.

v0.10 has some [experimental helpers](https://github.com/benallfree/pocketpages/blob/next/src/lib/db.ts#L11) because I found `model.Record` to be really annoying.

**Comment by bartonhammond** [comment]
Great - didn't realize that (model.record) Thanks!

My code is checked in if you want to look

I am now trying to deal with the "items left" - which is a count of the items that are not completed.

I thought to make a PB View but couldn't figure out the query.  The UI doesn't give very good diagnostics, imo. I wanted the query to summarize the counts grouped by "done".  Couldn't figure it out.  

So I'm trying to make a EJS function that I can call from multiple functions.  The code that is checked in has a "utils" in `_private/utils.js`.  Within the `/xapi/toggle` I include the 'utils' and call the function `getItemsLeft`.  But I get this:
<img width="1054" alt="Screenshot 2024-11-23 at 6 38 28 PM" src="https://github.com/user-attachments/assets/8edf95b8-41b5-434b-91a0-a426e0c6a054">

**Comment by benallfree** [comment]
Make sure you understand [EJS](https://ejs.co/#docs) and that you understand [partials](https://pocketpages.dev/docs/partials) vs [private files](https://pocketpages.dev/docs/private-files).

I think you mean to do `const { getItemsLeft } = requirePrivate('utils.ejs')`

**Reply by bartonhammond** [reply]
Well - thanks for all your help.  I find this to be a total mess - so many technologies mashed together and really confusing which technology is in play.  I give up.  Good luck.  But I can't proceed with this.  I thought that this PocketPages would be a breath of fresh air and bring some simplicity and clarity back but it appears to me to be a total mess.  Maybe if there was a comprehensive example it would help but I'm banging my head on the wall trying to figure out what context I'm in.  Maybe I'll come back later when there are more (ie any) examples.

**Comment by bartonhammond** [comment]
Done

**Comment by benallfree** [comment]
Thanks for trying it, sorry things are still somewhat raw. We'll keep working on it :)

It sounds like EJS and JSVM weren't familiar to you, is that right? I'm working on abstracting away both of those so it's "just PocketPages", but I agree it's not there yet.

**Comment by bartonhammond** [comment]
Let me know when you get this shaped up. I looked at a lot of different
technologies surrounding PocketBase and this was by far the most
interesting to me.  I really think this could be significant but it needs
some work.  I've always been an application developer so I don't have much
to say about how to make this work.  But I hope you continue.  I really
think all this context switching is a brain drain - kind of like what DHH
was talking about.  And the simplicity of this approach is somewhat like
what RoR is aiming for - no build steps, no complexity with the deployment
steps.  I hope debugging can be addressed too.  Going back to "log.info"
sucks - I was doing that 40 years ago.  If you could work w/ me more and
help figure this out, I would be agreeable.  Right now I'm banging my
head.  When I worked at IBM one of the roles was to write the example apps
that customers would use.  I enjoy that type of activity but I need
interaction and help to get through it.  I am not able to reverse engineer
this and all the various technologies are confusing - am I in EJS or BP and
in API or XAPI or Private or or Partials?  And depending on which, I have
different code.  Anyway, I'd love to work on this more but I'm exhausted
with my efforts.  :)

**Reply by benallfree** [reply]
Yeah I can feel the frustration for sure. Are you on Discord at all? If so, jump into the PocketHost channel https://pockethost.io/support and then we can chat live - much less frustrating than a message forum.

I think PocketPages is going to be a very important piece of the puzzle - it is for me already, but even I have trouble building apps with it. It's a back-and-forth thing: try something in the app, add something to PocketPages, back to the app, back to PocketPages.

**Comment by bartonhammond** [comment]
No thanks - don't like Discord at all.

**Comment by benallfree** [comment]
Ok thanks again for all your help and feedback. It has been super helpful for me and I think probably for others who follow, too. 'm leaving this issue open so it appears in the discussion area.

A couple points you helped clarify for me:

### JSVM Abstraction

The PocketBase JSVM API is like the assembly language of PocketBase. It's low-level, hard to use, and adds to the context switching confusion. I already started addressing this in v10 but I'm going to really commit to abstracting it away entirely. If there is something server-side to do, there should be a PocketPages API to do it. That way the complexity of the JSVM will be hidden away. People can still use it for very advanced cases, but most won't need to.

### No reserved API paths (xapi)

The whole `xapi` thing was confusing and I think, ultimately, unnecessary. PocketPages already forwards unmatched routes to PocketBase by simply passing them through the middleware. This may actually create an advantage where you can use PocketPages to protect `/_/*` (superadmin) routes and extend or add special cases to `/api/*` routes. 

### Clarifying `include` vs `require`

I think this is a good separation to keep. `include` expects an EJS template and returns a string. `require` loads and returns a JS module. But maybe `include` should be renamed `partial` or something like that.

**Comment by bartonhammond** [comment]
I am going to try and finish up this Todo app.  Only have a few more conditions to address.  

1).  Do you have an example of call a function in EJS.   I'd like to create a function that querys the data with different filters. Basically the code in `+load.js`
2).  The `+load.js` is confusing to me if using HTMX.  It's weird in having the `module.exports` stuff.  But if using HTMX should the data be loaded with a HTMX event?  
3)  Is there any way to attach a debugger?  I use VS Code but I'd switch if something else works.


https://github.com/bartonhammond/todomvc/blob/main/app/pages/_private/utils.ejs
Called here: https://github.com/bartonhammond/todomvc/blob/main/app/pages/xapi/toggle.ejs
Here's the error that occurs when the todo item "done" checkbox is clicked

```
PocketPages Error
GoError: /xapi/toggle.ejs:1
 >> 1| <%
    2| const { getItemsLeft } = requirePrivate('utils.ejs')
    3| 
    4| let formData = $apis.requestInfo(ctx).data

SyntaxError: /_private/utils.ejs: Line 1:38 Unexpected token < (and 3 more errors)
	at github.com/dop251/goja_nodejs/require.(*RequireModule).require-fm (native)
	at requirePrivate (node_modules/pocketpages/src/main.ts:304:16(9))
	at anonymous (<eval>:11:40(24))
	at apply (native)
	at anonymous (node_modules/pocketpages/node_modules/pocketbase-ejs/lib/ejs.js:702:16(24))
	at node_modules/pocketpages/node_modules/pocketbase-ejs/lib/ejs.js:490:27(103)
	at renderFile (node_modules/pocketpages/src/main.ts:387:28(24))
	at node_modules/pocketpages/src/main.ts:419:31(387)
	at apply (native)
	at <eval>:3:35(8)
```
